File: src/Core/Publication.fs
namespace ARCtrl

open ARCtrl.Helper
open Fable.Core

[<AttachMembers>]
type Publication(?pubMedID, ?doi, ?authors, ?title, ?status, ?comments) =

    let mutable _pubMedID : URI option = pubMedID
    let mutable _doi : string option = doi
    let mutable _authors : string option = authors
    let mutable _title : string option = title
    let mutable _status : OntologyAnnotation option = status
    let mutable _comments : ResizeArray<Comment> = Option.defaultValue (ResizeArray()) comments

    member this.PubMedID
        with get() = _pubMedID
        and set(pubMedID) = _pubMedID <- pubMedID

    member this.DOI
        with get() = _doi
        and set(doi) = _doi <- doi

    member this.Authors
        with get() = _authors
        and set(authors) = _authors <- authors

    member this.Title
        with get() = _title
        and set(title) = _title <- title

    member this.Status
        with get() = _status
        and set(status) = _status <- status

    member this.Comments
        with get() = _comments
        and set(comments) = _comments <- comments

    static member make pubMedID doi authors title status comments =
        Publication(?pubMedID=pubMedID, ?doi=doi, ?authors=authors, ?title=title, ?status=status, comments=comments)

    static member create(?pubMedID,?doi,?authors,?title,?status,?comments) : Publication =
       Publication.make pubMedID doi authors title status (Option.defaultValue (ResizeArray()) comments)

    static member empty() =
        Publication.create()

    member this.Copy() =
        let nextComments = this.Comments |> ResizeArray.map (fun c -> c.Copy())
        Publication.make this.PubMedID this.DOI this.Authors this.Title this.Status nextComments

    override this.GetHashCode() =
        [|
            HashCodes.boxHashOption this.DOI
            HashCodes.boxHashOption this.Title
            HashCodes.boxHashOption this.Authors
            HashCodes.boxHashOption this.PubMedID
            HashCodes.boxHashOption this.Status
            HashCodes.boxHashSeq this.Comments
        |]
        |> HashCodes.boxHashArray 
        |> fun x -> x :?> int

    override this.Equals(obj) =
        match obj with
        | :? Publication as p -> 
            this.GetHashCode() = p.GetHashCode()
        | _ -> false

    //let mutable _pubMedID : URI option = pubMedID
    //let mutable _doi : string option = doi
    //let mutable _authors : string option = authors
    //let mutable _title : string option = title
    //let mutable _status : OntologyAnnotation option = status
    //let mutable _comments : ResizeArray<Comment> = Option.defaultValue (ResizeArray()) comments

    override this.ToString() =
        let sb = System.Text.StringBuilder()
        sb.Append("Publication {\n\t") |> ignore
        [
            "PubMedID", this.PubMedID
            "DOI", this.DOI
            "Authors", this.Authors
            "Title", this.Title
            "Status", 
                this.Status
                |> Option.map (sprintf "%A")
            "Comments", 
                if Seq.length this.Comments > 0 then 
                    this.Comments
                    |> sprintf "%A"
                    |> Some
                else
                    None
        ] 
        |> List.choose (fun (s,opt) -> opt |> Option.map (fun o -> s,o))
        |> List.map (fun (s,v) -> sprintf "%s = %A" s v)
        |> String.concat ",\n\t"
        |> sb.Append
        |> ignore
        sb.Append("\n}") |> ignore
        sb.ToString()

File: js/Core/Publication.js
import { map as map_1, defaultArg, unwrap } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { ResizeArray_map } from "./Helper/Collections.js";
import { boxHashSeq, boxHashOption, boxHashArray } from "./Helper/HashCodes.js";
import { safeHash } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { StringBuilder__Append_Z721C83C5, StringBuilder_$ctor } from "../fable_modules/fable-library-js.4.22.0/System.Text.js";
import { printf, toText, join } from "../fable_modules/fable-library-js.4.22.0/String.js";
import { ofArray, choose, map } from "../fable_modules/fable-library-js.4.22.0/List.js";
import { length } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { toString } from "../fable_modules/fable-library-js.4.22.0/Types.js";
import { class_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class Publication {
    constructor(pubMedID, doi, authors, title, status, comments) {
        this._pubMedID = pubMedID;
        this._doi = doi;
        this._authors = authors;
        this._title = title;
        this._status = status;
        this._comments = defaultArg(comments, []);
    }
    get PubMedID() {
        const this$ = this;
        return unwrap(this$._pubMedID);
    }
    set PubMedID(pubMedID) {
        const this$ = this;
        this$._pubMedID = pubMedID;
    }
    get DOI() {
        const this$ = this;
        return unwrap(this$._doi);
    }
    set DOI(doi) {
        const this$ = this;
        this$._doi = doi;
    }
    get Authors() {
        const this$ = this;
        return unwrap(this$._authors);
    }
    set Authors(authors) {
        const this$ = this;
        this$._authors = authors;
    }
    get Title() {
        const this$ = this;
        return unwrap(this$._title);
    }
    set Title(title) {
        const this$ = this;
        this$._title = title;
    }
    get Status() {
        const this$ = this;
        return unwrap(this$._status);
    }
    set Status(status) {
        const this$ = this;
        this$._status = status;
    }
    get Comments() {
        const this$ = this;
        return this$._comments;
    }
    set Comments(comments) {
        const this$ = this;
        this$._comments = comments;
    }
    static make(pubMedID, doi, authors, title, status, comments) {
        return new Publication(unwrap(pubMedID), unwrap(doi), unwrap(authors), unwrap(title), unwrap(status), comments);
    }
    static create(pubMedID, doi, authors, title, status, comments) {
        const comments_1 = defaultArg(comments, []);
        return Publication.make(pubMedID, doi, authors, title, status, comments_1);
    }
    static empty() {
        return Publication.create();
    }
    Copy() {
        const this$ = this;
        const nextComments = ResizeArray_map((c) => c.Copy(), this$.Comments);
        const pubMedID = this$.PubMedID;
        const doi = this$.DOI;
        const authors = this$.Authors;
        const title = this$.Title;
        const status = this$.Status;
        return Publication.make(pubMedID, doi, authors, title, status, nextComments);
    }
    GetHashCode() {
        const this$ = this;
        return boxHashArray([boxHashOption(this$.DOI), boxHashOption(this$.Title), boxHashOption(this$.Authors), boxHashOption(this$.PubMedID), boxHashOption(this$.Status), boxHashSeq(this$.Comments)]) | 0;
    }
    Equals(obj) {
        let p;
        const this$ = this;
        return (obj instanceof Publication) && ((p = obj, safeHash(this$) === safeHash(p)));
    }
    toString() {
        let option, clo, arg_1;
        const this$ = this;
        const sb = StringBuilder_$ctor();
        StringBuilder__Append_Z721C83C5(sb, "Publication {\n\t");
        StringBuilder__Append_Z721C83C5(sb, join(",\n\t", map((tupledArg_1) => toText(printf("%s = %A"))(tupledArg_1[0])(tupledArg_1[1]), choose((tupledArg) => map_1((o) => [tupledArg[0], o], tupledArg[1]), ofArray([["PubMedID", this$.PubMedID], ["DOI", this$.DOI], ["Authors", this$.Authors], ["Title", this$.Title], ["Status", (option = this$.Status, map_1((clo = toText(printf("%A")), clo), option))], ["Comments", (length(this$.Comments) > 0) ? ((arg_1 = this$.Comments, toText(printf("%A"))(arg_1))) : undefined]])))));
        StringBuilder__Append_Z721C83C5(sb, "\n}");
        return toString(sb);
    }
}

export function Publication_$reflection() {
    return class_type("ARCtrl.Publication", undefined, Publication);
}

export function Publication_$ctor_Z645CED36(pubMedID, doi, authors, title, status, comments) {
    return new Publication(pubMedID, doi, authors, title, status, comments);
}

